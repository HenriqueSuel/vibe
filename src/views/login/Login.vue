<template>
  <div class="login">
    <h1 class="login__title">Entrar na minha conta</h1>

    <KindPerson
      :isPatient="isPatient"
      v-on:change-kind-person="isPatient = $event"
    ></KindPerson>

    <section class="card">
      <form @submit="submit" class="login__form">
        <input
          type="text"
          placeholder="CPF"
          v-model="$v.doc.$model"
          v-if="isPatient"
          v-mask="'###.###.###-##'"
        />
        <input
          type="text"
          placeholder="CRM"
          v-else
          v-model="$v.doc.$model"
          v-mask="'XXXXXXXXXXXXXXXXX'"
        />
        <span v-if="$v.doc.$error" :class="{ 'mensage-error': $v.doc.$error }"
          >{{ isPatient ? "CPF" : "CRM" }} Invalido</span
        >
        <button class="login__form__submit button-primary" type="submit">
          Entrar
        </button>
      </form>
    </section>
    <div class="login__register">
      <router-link to="/cadastro">Cadastra-se</router-link>
    </div>
  </div>
</template>

<script>
import { required, minLength } from 'vuelidate/lib/validators';
import KindPerson from '@/components/kind-person/KindPerson.vue';

export default {
  name: 'Login',
  components: {
    KindPerson,
  },
  data() {
    return {
      isPatient: false,
      doc: '',
    };
  },
  watch: {
    isPatient() {
      this.doc = '';
    },
  },
  computed: {
    rules() {
      return this.isPatient ? { minLength: minLength(14) } : {};
    },
  },
  validations() {
    return {
      doc: {
        required,
        ...this.rules,
      },
    };
  },
  methods: {
    async submit(e) {
      e.preventDefault();
      if (this.$v.doc.$error) return;
      const toast = {
        title: 'Usuario',
        description: 'Usuário não encontrado',
        type: 'error',
      };
      const user = this.isPatient
        ? await this.$store.dispatch(
          'PatientStore/actionPatientLogin',
          this.doc,
        )
        : await this.$store.dispatch('DoctorStore/actionDoctorLogin', this.doc);
      if (!user) this.$store.dispatch('GlobalStore/actionToast', toast);
    },
  },
};
</script>

<style lang="scss" scoped>
.login {
  padding: 16px;
  &__title {
    margin-top: 32px;
    text-align: center;
    color: #e72d51;
  }
  &__form {
    display: flex;
    flex-direction: column;
    &__type {
      margin-top: 8px;
      input {
        margin-right: 8px;
      }
    }
  }
  &__register {
    text-align: center;
    margin: 16px;
    a {
      text-decoration: none;
      color: black;
      &:hover {
        color: #e72d51;
      }
    }
  }
}
</style>
