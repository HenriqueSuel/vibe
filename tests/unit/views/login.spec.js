import { shallowMount, createLocalVue } from '@vue/test-utils';
import { VueMaskDirective } from 'v-mask';
import Vuex from 'vuex';
import flushPromises from 'flush-promises';

import Login from '@/views/login/Login.vue';
import Vuelidate from 'vuelidate';

const localVue = createLocalVue(Vuelidate);
localVue.use(Vuex);
localVue.use(Vuelidate);

describe('Login.vue', () => {
  let actions;
  let store;

  beforeEach(() => {
    actions = {
      actionPatientLogin: jest.fn().mockResolvedValue({ id: '1' }),
      actionDoctorLogin: jest.fn().mockResolvedValue({ id: '1' }),
      actionToast: jest.fn().mockResolvedValue(),
    };

    store = new Vuex.Store({
      state: {},
      modules: {
        PatientStore: {
          namespaced: true,
          actions,
        },
        DoctorStore: {
          namespaced: true,
          actions,
        },
        GlobalStore: {
          namespaced: true,
          actions,
        },
      },
    });
  });

  it('should change the input to cpf', () => {
    const wrapper = shallowMount(Login, {
      stubs: ['router-link'],
      directives: {
        mask: VueMaskDirective,
      },
      localVue,
      store,
    });

    wrapper.vm.$v.$touch();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should login in with the patient', async () => {
    const wrapper = shallowMount(Login, {
      stubs: ['router-link'],
      directives: {
        mask: VueMaskDirective,
      },
      localVue,
      store,
    });
    wrapper.setData({ isPatient: true, doc: '460.978.918-30' });
    const buttonSubmit = wrapper.find('button');
    buttonSubmit.trigger('submit');
    await flushPromises();
    expect(actions.actionPatientLogin).toHaveBeenCalled();
    expect(actions.actionToast).not.toHaveBeenCalled();
  });

  it('shouldnt login in with the patient', async () => {
    actions = {
      actionPatientLogin: jest.fn().mockResolvedValue(null),
      actionToast: jest.fn().mockResolvedValue(),
    };

    store = new Vuex.Store({
      state: {},
      modules: {
        PatientStore: {
          namespaced: true,
          actions,
        },
        GlobalStore: {
          namespaced: true,
          actions,
        },
      },
    });
    const wrapper = shallowMount(Login, {
      stubs: ['router-link'],
      directives: {
        mask: VueMaskDirective,
      },
      localVue,
      store,
    });
    wrapper.setData({ isPatient: true, doc: '460.978.918-30' });
    const buttonSubmit = wrapper.find('button');
    buttonSubmit.trigger('submit');
    await flushPromises();
    expect(actions.actionPatientLogin).toHaveBeenCalled();
    expect(actions.actionToast).toHaveBeenCalled();
  });

  it('After the user submit button, $v.$invalid return true?', () => {
    const wrapper = shallowMount(Login, {
      stubs: ['router-link'],
      directives: {
        mask: VueMaskDirective,
      },
      localVue,
      store,
    });
    const loginBtn = wrapper.find('button');
    loginBtn.trigger('submit');
    expect(wrapper.vm.$v.$invalid).toBeTruthy();
  });

  it('should log in with the doctor', async () => {
    const wrapper = shallowMount(Login, {
      stubs: ['router-link'],
      directives: {
        mask: VueMaskDirective,
      },
      localVue,
      store,
    });
    wrapper.setData({ isPatient: false, doc: '46097891830' });
    const buttonSubmit = wrapper.find('button');
    buttonSubmit.trigger('submit');
    await flushPromises();
    expect(actions.actionDoctorLogin).toHaveBeenCalled();
  });
});
