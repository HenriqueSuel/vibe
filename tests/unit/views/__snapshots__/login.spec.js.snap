// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login.vue should change the input to cpf 1`] = `
<div class="login">
  <h1 class="login__title">Entrar na minha conta</h1>
  <kindperson-stub></kindperson-stub>
  <section class="card">
    <form class="login__form"><input type="text" placeholder="CRM">
      <!----> <button type="submit" class="login__form__submit button-primary">
        Entrar
      </button>
    </form>
  </section>
  <div class="login__register">
    <router-link-stub to="/cadastro">Cadastra-se</router-link-stub>
  </div>
</div>
`;
